Expanded teams table to handle delegates - Fixes #680

pingskills opened the pull request:
The new structure has committees, each committee can have
multiple teams, and each team has multiple team members.

Committees have committee positions which are assigned to
team members. There is a team leader flag on each committee position.

For delegates each senior delegate has their own team. The
delegate committee is the only committee initially set up
with multiple teams but I could see the software team using
multiple teams in the future.

This was a much bigger change than I had anticipated :) I'd love to hear your thoughts. Is this something to consider implementing or is it far too complex?

pingskills commented:
Here are some screenshots:

![editing-team](https://cloud.githubusercontent.com/assets/4403168/16104006/cecd5898-33c0-11e6-822f-12deace0e2a5.png)
![editing-team-member](https://cloud.githubusercontent.com/assets/4403168/16104008/cecf2eca-33c0-11e6-9ae7-8869a4ed5853.png)
![delegates-committee-page](https://cloud.githubusercontent.com/assets/4403168/16104010/ced2a06e-33c0-11e6-9220-26322888df96.png)
![committees-page](https://cloud.githubusercontent.com/assets/4403168/16104007/cece31a0-33c0-11e6-9776-755550e1c120.png)
![wca-id-claim-email](https://cloud.githubusercontent.com/assets/4403168/16104009/cecfbc32-33c0-11e6-9ce9-a902d868b714.png)

jonatanklosko commented on WcaOnRails/app/controllers/committee_positions_controller.rb:
```diff
@@ -0,0 +1,72 @@
+class CommitteePositionsController < ApplicationController
+  before_action :authenticate_user!, except: [ :index, :show ]
```
Why do you have `except` here? There are no index and show actions.

pingskills replied:
Good point. I'll remove the except.

jonatanklosko commented on WcaOnRails/app/controllers/committee_positions_controller.rb:
```diff
@@ -0,0 +1,72 @@
+  before_action -> { redirect_unless_user(:can_manage_committees?) }
+
+  def index
+    @committee = Committee.find_by_slug(params[:committee_id])
```
Following our convention there should be `committee_from_params` =)

jonatanklosko commented on WcaOnRails/app/controllers/committee_positions_controller.rb:
```diff
@@ -0,0 +1,72 @@
+    CommitteePosition.find_by_committee_id_and_slug(committee_id, params[:id])
+  end
+
+  private def committee_position_from_params()
```
These parentheses are unnecessary.

jonatanklosko commented on WcaOnRails/app/controllers/committee_positions_controller.rb:
```diff
@@ -0,0 +1,72 @@
+
+  def destroy
+    @committee = Committee.find(params[:committee_id])
+    @committee_position = committee_position_from_params()
```
Unnecessary parentheses here as well.

jonatanklosko replied:
Actually there are a lot of such places, but I think rubocop will inform you about the. Do you run it locally?

pingskills replied:
No I haven't run rubocop locally. I'll look at how to get that set up.

jonatanklosko replied:
As simpls as `bundle exec rubocop` ^^

jonatanklosko commented on WcaOnRails/app/controllers/committee_positions_controller.rb:
```diff
@@ -0,0 +1,72 @@
+      flash[:success] = "Created new committee position:" + @committee_position.name
+      redirect_to committee_committee_positions_path(@committee.slug)
+    else
+      flash[:danger] = "Failed to create committee position:" + @committee_position.name
```
Since you show error messages this is unmecessary. Actually this has been discussed recently.

jonatanklosko replied:
I'm commenting here but the same goes for other lines like this one of course.

pingskills replied:
OK. So I'll remove the flash messages.

jonatanklosko commented on WcaOnRails/app/controllers/committee_positions_controller.rb:
```diff
@@ -0,0 +1,72 @@
+  end
+
+  def committee_position_params
+    return committee_position_params = params.require(:committee_position).permit(:name, :description, :team_leader)
```
Both `return` and the variable is unnecessary here.

jonatanklosko commented on WcaOnRails/app/controllers/committee_positions_controller.rb:
```diff
@@ -0,0 +1,72 @@
+  end
+
+  def new
+    @committee = Committee.find_by_slug(params[:committee_id])
```
It's very inconsistent to use both `id` `slug` . The solution is to override `to_param` method for Committee to return the slug. This will allow you to
- use one `committee_from_params`
- don't pass slug to the url helpers; `committee_path(@committee)` will use slug

Note: the slug will still be in `params[:id]`.
I guess the same goes for other controllers

pingskills replied:
That sounds much better. For the committee_position I might revert to using the id, since I wanted the slug not to be unique. That way 2 different committees could have a team-leader slug. For the teams, and committees I think this will clean things up a lot.

jonatanklosko commented on WcaOnRails/config/routes.rb:
```diff
@@ -58,7 +58,13 @@
+    resources :teams, only: [:new, :edit, :create, :update, :destroy] do
+      resources :team_members, only: [:new, :edit, :create, :update]
+    end
+    resources :committee_positions, only: [:index, :new, :edit, :create, :update, :destroy]
```
This leads to weird url helpers (`committee_committee_positions_path`) and urls (`/committee/committee_positions`).
The solution is pretty simple: add two more options
`as: :positions` (for url helpers) and `paths: :positions` (for urls).

pingskills replied:
Good suggestion. That makes it easier to read. Perhaps I should of called the model positions.

pingskills replied:
This messes up my simple_form.

`simple_form_for [@committee, @position]`

This tries to find committee_committee_path but if I pass in the URL option, it is not the same for the edit and the new. Is there a way to deal with this?

jonatanklosko replied:
Ah I see. Actually I've stumbled upon identical issue in my most recent PR. The way I managed to fix it was to pass variale `url` to the form partial and then use it with simple form `url: url`.

jonatanklosko commented on WcaOnRails/app/controllers/committees_controller.rb:
```diff
@@ -0,0 +1,58 @@
+  end
+
+  def committee_params
+    return committee_params = params.require(:committee).permit(:name, :email, :duties)
```
Unnecessary `return` and assignment.

jonatanklosko commented on WcaOnRails/app/controllers/teams_controller.rb:
```diff
@@ -1,53 +1,69 @@
       end
+    else
+      flash[:error] = "Cannot delete a team whilst it still has team members. If you really want to delete this team then delete the team members first."
+      redirect_to committee_path(@committee.slug)
```
In each case you perform the same redirection, so it can be moved out of the `if` statement.

jonatanklosko commented on WcaOnRails/app/helpers/static_pages_helper.rb:
```diff
@@ -1,7 +1,7 @@
-  def format_team_members(team_friendly_id)
-    Team.find_by_friendly_id!(team_friendly_id).current_members.includes(:user).order(team_leader: :desc).order("users.name asc").map do |u|
-      u.user.name + (u.team_leader ? " (leader)" : "")
+  def format_team_members(slug)
```
The name is no longer accurate =)

pingskills replied:
It still is as this is actually getting the members for a team, not a committee.

jonatanklosko commented on WcaOnRails/app/models/committee.rb:
```diff
@@ -0,0 +1,22 @@
+  private def compute_slug
+    if !self.slug
+      self.slug = name.parameterize
+    end
```
Let's brief it
```ruby
self.slug ||= name.parameterize
```

jonatanklosko commented on WcaOnRails/app/models/committee_position.rb:
```diff
@@ -0,0 +1,21 @@
+  private def compute_slug
+    if !self.slug
+      self.slug = name.parameterize
+    end
```
This can be briefed as above.

jonatanklosko commented on WcaOnRails/app/models/team.rb:
```diff
@@ -1,21 +1,21 @@
+  private def compute_slug
+    if !self.slug
+      self.slug = name.parameterize
     end
```
Same as above.

jonatanklosko commented on WcaOnRails/app/models/team_member.rb:
```diff
@@ -13,7 +15,25 @@ def current_member?
+          errors.add(:user_id, "must not have overlapping membership dates with the #{self.team.name}.")
+        end
+      end
     end
```
This *must* be in the Team model. Consider creating a new member:
In the Team model `team_members` contains the new members (which aren't created) so the validatition works well.
Here you retreive `team_members` from database so the new ones are lost.

pingskills replied:
Hi @jonatanklosko,

I'm not sure I follow here. This validation is not for all members, but for a single member.

The new record which isn't saved yet has it's dates put into the `new_period` variable. Then I retrieve all other records for this `user_id` except the record this is updating (if it's an update where self.id exists). So the only new record is the one stored in self?

Let me know if that makes sense or if I am not quite understanding the process here.

jonatanklosko replied:
I was under assumption that there may be many new team members at once (this was possible before). Nice that this can be simplified now. I'd change it a bit:
```ruby
def membership_periods_cannot_overlap_for_single_user
  if start_date
    self_period = start_date..(end_date || Date::Infinity.new)
    other_members = team.team_members.where(user_id: user_id).where.not(id: id)
  
    other_members.each do |team_member|
      other_period = team_member.start_date..(team_member.end_date || Date::Infinity.new)
      if self_period.overlaps? other_period
          errors.add(:user_id, "must not have overlapping membership dates within #{self.team.name}.")
      end
    end
  end
end
```
Note that even if the id is nil this evaluates to `where.not(id: nil)` which should work.

pingskills replied:
Nice. I like that.

jonatanklosko commented on WcaOnRails/app/models/team_member.rb:
```diff
@@ -24,6 +44,34 @@ def cannot_demote_oneself
+      unless self.committee_position.slug == 'senior-delegate' && self.current_member?
+        senior_delegate_count = 0;
+        team_members = self.team.current_members
+        team_members.each do |team_member|
```
The variable doesn't seem to clean things, thiss would be better I think:
```ruby
team.current_members.each do ...
```

jonatanklosko commented on WcaOnRails/app/models/team_member.rb:
```diff
@@ -24,6 +44,34 @@ def cannot_demote_oneself
+  validate :delegate_team_must_have_at_least_one_senior_delegate
+  def delegate_team_must_have_at_least_one_senior_delegate
+    if self.team.committee.slug == "wca-delegates-committee"
+      unless self.committee_position.slug == 'senior-delegate' && self.current_member?
```
`unless` is harder to read and could be missleading when used with logical AND. (If this seems to be only my personal feeling then feel free to ignore this =) ).

pingskills replied:
I'll update this to `if !self.user.senior_delegate?` which checks if the user is currently a senior delegate.

jonatanklosko replied:
Neat =)

pingskills replied:
HI @jonatanklosko, unfortunately this didn't work :( 

The problem is that self.user.senior_delegate? expects the record to already be saved in the database. As we are validating the team_member this doesn't work. I'll switch it to an if statement.

pingskills replied:
Actually I'll put a senior_delegate? method in the `TeamMember` model and use that.

jonatanklosko commented on WcaOnRails/app/models/user.rb:
```diff
@@ -10,15 +12,23 @@ class User < ActiveRecord::Base
+  WCA_SOFTWARE_COMMITTEE_SLUG = "wca-software-committee".freeze
+  WCA_RESULTS_COMMITTEE_SLUG = "wca-results-committee".freeze
+  WCA_REGULATIONS_COMMITTEE_SLUG = "wca-regulations-committee".freeze
+  WCA_DISCIPLINARY_COMMITTEE_SLUG = "wca-disciplinary-committee".freeze
```
These constants should go committee.rb I think.

jonatanklosko commented on WcaOnRails/app/models/user.rb:
```diff
@@ -302,74 +264,95 @@ def preferred_events
-    delegate_status.present?
+  def admin?
+    software_committee?
   end
```
```ruby
alias_method :admin?, :software_committee?
```
Would be nicer =)

jonatanklosko commented on WcaOnRails/app/models/user.rb:
```diff
@@ -530,24 +508,31 @@ def approve_pending_avatar!
 
       if ActiveRecord::Type::Boolean.new.type_cast_from_database(params[:only_delegates])
-        users = users.where.not(delegate_status: nil)
+        users = users.joins(:current_committees).where(committees: {slug: WCA_DELEGATES_COMMITTEE_SLUG})
```
I'd suggest a `scope :delegates` for this =)

jonatanklosko commented on WcaOnRails/app/views/committee_positions/_committee_position_form.html.erb:
```diff
@@ -0,0 +1,16 @@
+    <%= render 'shared/error_messages', object: f.object %>
+
+    <%= f.input :name %>
+    <%= f.input :description %>
```
Markdown editor should be used here since you render description with `md`.

jonatanklosko commented on WcaOnRails/app/views/committee_positions/index.html.erb:
```diff
@@ -0,0 +1,22 @@
+  <% if @committee_positions.empty? %>
+    <%= alert :warning, "There are currently no committee_positions." %>
+  <% else %>
+    <% @committee_positions.includes(:team_members).each do |position| %>
```
Loose suggestion: what about using trash icon? =)

jonatanklosko replied:
Oh, the comment should go below where icon "remove" is used.

jonatanklosko commented on WcaOnRails/app/views/committees/_committee_form.html.erb:
```diff
@@ -0,0 +1,21 @@
+  boolean: :horizontal_boolean
+} do |f| %>
+  <%= render 'shared/error_messages', object: f.object %>
+  <%= f.input :name, hint: t('.name_html') %>
```
Hints have their own place in the en.yml file and should go right there =)

jonatanklosko commented on WcaOnRails/app/views/committees/show.html.erb:
```diff
@@ -0,0 +1,76 @@
+      <% if @committee.teams.empty? %>
+        <%= link_to icon("remove", "Delete Committee"), committee_path(@committee), method: :delete, data: { confirm: 'Are you sure you want to delete this committee?' }, class: "btn btn-danger" %>
+      <% end %>
+      <%= link_to icon("new", "Create New Team"), new_committee_team_path(@committee.slug), class: "btn btn-info" %>
```
This doesn't seem to work on the screenshot. Actually new isn't thw actual fontawesome icon. Let's use "plus" then. (That's not the only one place using this)

jonatanklosko commented on WcaOnRails/app/views/committees/show.html.erb:
```diff
@@ -0,0 +1,76 @@
+
+<div class="container">
+  <h1><%= yield(:title) %></h1>
+  <p><a href="mailto:<%= @committee.email %>"><%= @committee.email %></a></p>
```
Rails helper will help to clean this up:
```ruby
<%= mail_to @committee.email %>
```

jonatanklosko commented on WcaOnRails/app/views/teams/_team_form.html.erb:
```diff
@@ -0,0 +1,15 @@
+    <%= render 'shared/error_messages', object: f.object %>
+
+    <%= f.input :name %>
+    <%= f.input :description %>
```
What about using the markdown editor here? (If you will do so then remember to render it as markdown as well).

pingskills replied:
Good idea. I'll update this.

jonatanklosko commented on WcaOnRails/app/views/committees/show.html.erb:
```diff
@@ -0,0 +1,76 @@
+        <%= link_to icon("edit", "Edit #{team.name}"), edit_committee_team_path(@committee.slug, team.slug), class: "btn btn-info" %>
+        <%= link_to icon("user-plus", "Add a member to #{team.name}"), new_committee_team_team_member_path(@committee.slug, team.slug), class: "btn btn-info" %>
+        <% if team.team_members.empty? %>
+          <%= link_to icon("remove", "Delete #{team.name}"), committee_team_path(@committee, team), method: :delete, data: { confirm: 'Are you sure you want to delete this team?' }, class: "btn btn-danger" %>
```
Loose suggestion: "trash" here as well =)

jonatanklosko commented on WcaOnRails/app/views/committees/show.html.erb:
```diff
@@ -0,0 +1,76 @@
+    <% if current_user&.can_manage_committees? %>
+      <%= link_to icon("edit", "Edit Committee"), edit_committee_path, class: "btn btn-info" %>
+      <% if @committee.teams.empty? %>
+        <%= link_to icon("remove", "Delete Committee"), committee_path(@committee), method: :delete, data: { confirm: 'Are you sure you want to delete this committee?' }, class: "btn btn-danger" %>
```
Please split this into two lines.

Same as above.

jonatanklosko commented on WcaOnRails/app/models/user.rb:
```diff
@@ -302,74 +264,95 @@ def preferred_events
-  def software_team?
-    team_member?('software')
+  def delegate_status
+    committee_positions = current_team_members.joins(:committee).where(committees: {slug: 'wca-delegates-committee'}).map(&:committee_position)
```
We should be consistent and use the constant slugs *everywhere*.

jonatanklosko commented on WcaOnRails/app/views/static_pages/contact.html.erb:
```diff
@@ -12,38 +12,15 @@
+  <h2><u><%= link_to "Committees and Teams", committees_path %></u></h2>
+  <% @committees.each do |committee| %>
+    <p>
+      <strong><%= link_to committee.name, committee_path(committee.slug) %></strong> - <a href="mailto:<%= committee.email %>"><%= committee.email %></a><br />
```
Let's use `mail_to` here as well.

jonatanklosko commented on WcaOnRails/app/views/team_members/_team_member_form.html.erb:
```diff
@@ -0,0 +1,21 @@
+    <% else %>
+      <%= f.input :user_id, required: true, as: :user_ids, only_one: true %>
+      <%= f.association :committee_position, collection: @committee.committee_positions, label_method: :name, value_method: :id %>
+    <% end %>
```
What about having two inputs here and using
`disabled: @team_member.persisted?`

jonatanklosko replied:
You either missed it or I wasn't clear, here's what I thought
```erb
  <%= f.input :user_id, required: true, as: :user_ids, only_one: true, disabled: @team_member.persisted? %>
  <%= f.association :committee_position, collection: @committee.committee_positions, label_method: :name, 
                                         value_method: :id, disabled: @team_member.persisted? %>
```

pingskills replied:
That works nicely :)

jonatanklosko commented on WcaOnRails/config/locales/en.yml:
```diff
@@ -405,3 +405,16 @@ en:
+    new:
+      create_committee: "Create New Committee"
+    committee_form:
+      supports_md_html: "Supports <a href='https://daringfireball.net/projects/markdown/basics' target='_blank'>Markdown</a>"
```
@viroulep do you know how to avoid duplications like this one? Is there a way to make an alias or something? Repeating will be painful for translators.

viroulep replied:
I attempted to create a "common" section in the locale file, for words that appeared often and would be translated the same way regardless of the context.
Right now the md description is always used as a parameter for a [hint](https://github.com/pingskills/worldcubeassociation.org/blob/3d852accfdcb2a3eef6939a401eca2b5e7b0598e/WcaOnRails/app/views/competitions/_competition_form.html.erb#L41), which means we always have to specify the key. My suggestion would be to move this common hint to `common.hints.supports_md_html`, and use this key in every forms (which also implies removing the `competition.competition_form.supports_md_html`).

jonatanklosko commented on WcaOnRails/spec/controllers/api_controller_spec.rb:
```diff
@@ -403,19 +389,15 @@ def api_sign_in_as(user, scopes: nil)
-
-        wrc_team = Team.find_by_friendly_id('wrc')
-        FactoryGirl.create(:team_member, team_id: wrc_team.id, user_id: user.id)
+        user = FactoryGirl.create :regulations_team
```
Looks like you assign a team to the `user` variable.

pingskills replied:
I'll update this to be `FactoryGirl.create :regulations_team_member`

jonatanklosko commented on WcaOnRails/spec/controllers/committee_positions_controller_spec.rb:
```diff
@@ -0,0 +1,187 @@
+        get :show, {committee_id: committee_position.committee.slug}
+      }.to raise_error(ActionController::UrlGenerationError)
+    end
+  end
```
I feel like it doesn't make sense to test actions that doesn't exist.

jonatanklosko replied:
There are more tests like this one, the comment applies to all of them.

pingskills replied:
Maybe not but I wanted to ensure that it's obvious these default routes shouldn't be used. This way if someone adds the route they will have to remove the test.

jonatanklosko replied:
Sorry, but I don't see how this could be valuable. For me it's almost like testing whether `GET #foo` raises the error. But maybe @larspetrus have another opinion =) ?

pingskills replied:
Cool. I'll get rid of these tests.

jonatanklosko commented on WcaOnRails/spec/controllers/committee_positions_controller_spec.rb:
```diff
@@ -0,0 +1,187 @@
+      sign_out
+
+      it 'redirects to the sign in page' do
+        get :new, {committee_id: committee_position.committee.slug}
```
You don't need explicit hash here. Same for other tests.

jonatanklosko commented on WcaOnRails/spec/controllers/team_members_controller_spec.rb:
```diff
@@ -0,0 +1,178 @@
+#        user = FactoryGirl.create(:user)
+#        FactoryGirl.create(:team_member, team_id: team_leaders_team.id, user_id: user.id)
+#        user
+#      end
```
A comment is left =)

jonatanklosko commented on WcaOnRails/spec/controllers/team_members_controller_spec.rb:
```diff
@@ -0,0 +1,178 @@
+#        expect(team.slug).to eq original_slug
+#      end
+#    end
+#  end
```
Another comment =)

pingskills replied:
I need to add some tests here for POST#update

jonatanklosko commented on WcaOnRails/spec/factories/committee_positions.rb:
```diff
@@ -0,0 +1,9 @@
+FactoryGirl.define do
+  factory :committee_position do
+    sequence(:name) { |n| "Position #{n}" }
+    slug { name.parameterize }
```
Slug should be computed automatically in the model, right?

pingskills replied:
Good point. Only after create not build. And I guess that's how it happens so I'll take this out of the factory.

jonatanklosko commented on WcaOnRails/spec/factories/team_members.rb:
```diff
@@ -1,8 +1,90 @@
     end_date nil
+    after(:build) do |team_member|
+      committee_position = FactoryGirl.create(:committee_position, committee_id: team_member.team.committee_id)
+      team_member.committee_position_id = committee_position.id
```
I think this assignment doesn't update database.

jonatanklosko replied:
Yeah, I've just checked and such an assignment just sets the attribute but when you reload an object the value disappears. The best solution is to do
```ruby
team_member.committee_position = committee_position
```
as it immediately produces SQL query that updates the database.

pingskills replied:
Hi @jonatanklosko,

If I do the following:
`tm = FactoryGirl.build(:team_member)` it doesn't save `tm` to the database so I can't reload.

If I do:

`tm = FactoryGirl.create(:team_member)` then it saves the record so then the assignment is saved. So a reload will work. So I think this is OK? Let me know if this isn't right.

jonatanklosko replied:
Ah got it, it's after build whereas I've read after create.

jonatanklosko commented on WcaOnRails/spec/factories/team_members.rb:
```diff
@@ -1,8 +1,90 @@
+
+    trait :candidate_delegate do
+      after(:build) do |team_member|
+        candidate_delegate_position = CommitteePosition.joins(:committee).where(committees: {slug: 'wca-delegates-committee'}).find_by_slug('candidate-delegate')
```
Does this `position` exist in the database in the test environment?

jonatanklosko replied:
Sorry, I've just reached the end of the diff and I see it's created.

jonatanklosko commented on WcaOnRails/spec/features/create_committee_spec.rb:
```diff
@@ -0,0 +1,23 @@
+
+    new_committee = Committee.find_by_slug("random-ideas-committee")
+    expect(new_committee.name).to eq "Random Ideas Committee"
+    expect(new_committee.duties).to eq "Come up with random ideas about cubing."
```
As it's feature spec better approach would be to go to committees index and make expectations on the page content rather than checking database.

jonatanklosko commented on WcaOnRails/spec/support/database_cleaner.rb:
```diff
@@ -13,10 +13,32 @@
+    FactoryGirl.create(:team, slug: 'software-team', name: 'Software Team', description: 'Does software', committee_id: @software.id)
+    FactoryGirl.create(:team, slug: 'results-team', name: 'Results Team', description: 'Posts results', committee_id: @results.id)
+    FactoryGirl.create(:team, slug: 'regulations-team', name: 'Regulations Team', description: 'Regulations', committee_id: @regulations.id)
+    FactoryGirl.create(:team, slug: 'disciplinary-team', name: 'Disciplinary Team', description: 'Disciplinary Committee', committee_id: @disciplinary.id)
```
This seems a bit scary to me. Isn't this going to make our tests noticeably slower?

pingskills replied:
We currently create 4 teams, so there certainly is extra data being created.

jonatanklosko commented:
[Teams seeds](https://github.com/cubing/worldcubeassociation.org/blob/master/WcaOnRails/db/seeds/teams.seeds.rb) should be updated with all the data.

jonatanklosko commented:
A global note: I see that in the migrations you use SQL to insert records. My suggestion is to use Rails to do this. I'm sure this will make the migrations cleaner and more readable.
You can create multiple objects at once, like:
```ruby

Committee.create([ {
    name: 'WCA Board',
    slug: 'wca-board',
    duties: 'Responsible for the governance of the WCA.',
    email: 'board@worldcubeassociation.org',
  }, {
    name: 'WCA Disciplinary Committee',
    slug: 'wca-disciplinary-committee',
    duties: 'This committee advises the WCA Board in special cases, like alleged violations of WCA Regulations, and may be contacted by WCA members in case of important personal matters regarding WCA competitions.',
    email: 'wdc@worldcubeassociation.org',
  },
  # ...
]);
```

jonatanklosko commented on WcaOnRails/app/views/committee_positions/_committee_position_form.html.erb:
```diff
@@ -0,0 +1,16 @@
+    <%= f.input :name %>
+    <%= f.input :description %>
+    <%= f.input :team_leader %>
+    <%= link_to icon("undo", "Back to #{@committee.slug}"), committee_path(@committee.slug), class: "btn btn-primary" %>
```
Should be "Back to #{@committee.name}".

jonatanklosko commented on WcaOnRails/app/views/team_members/_team_member_form.html.erb:
```diff
@@ -0,0 +1,21 @@
+    <% end %>
+    <%= f.input :start_date, as: :date_picker %>
+    <%= f.input :end_date, as: :date_picker %>
+    <%= link_to icon("undo", "Back to #{@committee.slug}"), committee_path(@committee.slug), class: "btn btn-primary teams-home" %>
```
Should be "Back to #{@committee.name}"

jonatanklosko commented on WcaOnRails/app/views/teams/_team_form.html.erb:
```diff
@@ -0,0 +1,15 @@
+
+    <%= f.input :name %>
+    <%= f.input :description %>
+    <%= link_to icon("undo", "Back to #{@committee.slug}"), committee_path(@committee.slug), class: "btn btn-primary teams-home" %>
```
Should be "Back to #{@committee.name}"

larspetrus commented:
> A global note: I see that in the migrations you use SQL to insert records. My suggestion is to use Rails to do this. I'm sure this will make the migrations cleaner and more readable.

This is actually a bit controversial.

If you ever rerun old migrations, the Model classes can easily have changed, making your migration produce different results and/or crash. So I've been taught to always use SQL in migrations.

If we always use migrations right away and never again in this project, using Model classes is fine, I guess. Still makes me nervous...

jonatanklosko commented:
@larspetrus ah that's a good note. I haven't taken this into account, thanks.

viroulep commented on WcaOnRails/config/locales/en.yml:
```diff
@@ -405,3 +405,16 @@ en:
+      name_html: "The name of the committee (e.g., Delegates, or Software)"
+      email_html: "The email of the committee (e.g., software@worldcubeassociation.org)"
+      slug_html: "This will be used as part of the public URL for this committee."
+      duties_html: "What this committee is responsible for? %{md}"
```
As @jonatanklosko said these should be in `simple_form.hints`, but this last one will probably stay here.
Also note that the suffix `_html` just tells rails that the text is html safe and should not be escaped, if the text doesn't contain html you can drop it :)

pingskills replied:
Thanks @viroulep, I'll move the supports_md_html into common.hints and move these others into hints without the html suffix.

larspetrus commented:
> This was a much bigger change than I had anticipated :) I'd love to hear your thoughts. Is this something to consider implementing or is it far too complex?

I'm not clear what real world problems are solved by this. This might be overkill, but since I don't know what we intend to kill, I can't tell. #680 is quite vague.

Top level design comments:

1. A `Committee` seems to just be a `Team` with a few extra capabilities. If you instead added those capabilities to `Team`, I think a lot of things would be simpler.

2. Is `CommitteePosition` doing more than giving committee members a title? It seems like that could be accomplished by just adding a `title` field to `TeamMember`.

pingskills commented:
Hi @larspetrus,

My understanding of the requirement is to be able to track the history of delegates. The current teams table stores the history of each team member but as each delegate also has a delegate status it doesn't fit straight into the current `TeamMember`.

I created `Committee` to be able to group teams. This information could be put on each team but then those attributes are duplicated.

A `CommitteePosition` is essentially there to give members a title but can also contain extra information like whether the position represents a team leader.

jfly commented:
One reason we want to track delegate history is in order to fix our validations (see https://github.com/cubing/worldcubeassociation.org/issues/400). Right now, we require that competition delegates are delegates, but what we really want is to require that they were delegates at the time of the competition. I haven't looked at this diff yet, but hopefully Jeff dealt with that in this PR.

larspetrus commented:
OK, tracking delegate history is a good concrete feature. That makes things clearer.

It looks like the way teams are grouped into committees is by the `committee_id` column in `teams`. My thought is that if committees were modeled as Teams, that could instead be a `parent_team_id`. That also makes an arbitrarily deep hierarchy possible.

This assumes that a team hierarchy is needed. I don't know if this is true. Can you not keep track of delegate history without it?

I can't quite make out how individual membership in committees is modeled. Looks like some combination of `committee_position` and `team_member`? Anyway, if committees are teams, this can be done with the existing `team_members` mechanism. Team_members already have `start_date` and `end_date`, so for delegate history purposes this should be all we need.

I understand the need to track properties of the team membership, like what kind of delegate you are, if you're the leader, etc. I think that is far simpler done by storing that in a `team_membership` column. The "tag" pattern I mentioned in some earlier discussion is one possible implementation.

If I'm right, this would simplify a _lot_. If I'm wrong, which is _very_ possible, let's talk about how.

pingskills commented:
Having a parent_id on teams is different to having a committee grouping together teams. Since there are many "delegate" type teams I thought it was a good idea to create a wrapper above them, hence the committees.

Perhaps a team hierarchy is not needed for delegates. You could certainly track delegate history by other means but this does seem reasonable since we already do it for other teams.

Users aren't individual members of committees (since these are just groupings of teams) but rather team members of a team. The start date and end date for team members is what tracks the history of each team membership.

As for tracking team membership properties, I like to do this in a separate table rather than a `team_membership` column, hence why I developed it this way. And because each team belongs to a committee, these properties are stored once per committee in committee_positions rather than once for each team.

pingskills commented:
I've just updated the code based on all the feedback from @jonatanklosko and @viroulep. 

Looks like we still need a lot more discussion on how we wish to implement delegate history though :) Either way we still need to decide how to handle regions. One suggestion was to put a longitude and latitude column on to the users table.

pingskills commented:
I've now updated the validation of competitions so that it checks the effective date of a delegate to ensure they were a delegate at the start date of the competition. If the competition does not have a start date yet then the delegate must be current.

In the migration all delegates get a start date of the first competition they delegates. Any delegate who is no longer a delegate gets and end_date of the last competition they delegated.

I've cleaned up the test cases so that the `database_cleaner.rb` does not need to create as much data.

@jonatanklosko - I wouldn't mind putting this on staging at some stage for people to have a play with. Let me know if you think that's a good idea, and if so can you give me some help on how to do that.

pingskills commented:
I've updated the PR based on feedback from when this change was on staging. I've made the following changes:

- used different colours for different buttons
- only show private email addresses for delegates
- only show the title of each team and not the description

@pedrosino suggested we also keep the country information which was captured on the users table but is not currently in the team_members table. I'd love to hear some thoughts on how to handle this. One way would be to tackle issue #732 and use the latitude and longitude for a user to display some information. Another would be to add an additional field to record some comments. Note that this applies to all teams so we could make it a generic field. Here's what 2 different teams currently look like:

![delegates](https://cloud.githubusercontent.com/assets/4403168/18155728/bc6c3b22-7053-11e6-897f-7fe9d096f595.png)

![software](https://cloud.githubusercontent.com/assets/4403168/18155729/c347a468-7053-11e6-9ea2-06743009fc2d.png)

Let me know your thoughts about this or anything else to do with the PR.

jonatanklosko commented on WcaOnRails/app/controllers/committee_positions_controller.rb:
```diff
@@ -0,0 +1,69 @@
+
+  def committee_from_params
+    Committee.find_by_slug(params[:committee_id])
+  end
```
These two should be private.

jonatanklosko commented on WcaOnRails/app/controllers/committees_controller.rb:
```diff
@@ -0,0 +1,67 @@
+
+  def to_params
+    slug
+  end
```
Shouldn't this be in the model?

pingskills replied:
I'm not sure what you mean here @jonatanklosko?

pingskills replied:
Ah yes, it is actually in the model (def to_param). I'm not sure why it's here. I'll get rid of it.

jonatanklosko commented on WcaOnRails/app/controllers/committees_controller.rb:
```diff
@@ -0,0 +1,67 @@
+
+  def committee_params
+    params.require(:committee).permit(:name, :email, :duties)
+  end
```
These two should be private.

jonatanklosko commented on WcaOnRails/app/controllers/team_members_controller.rb:
```diff
@@ -0,0 +1,63 @@
+  before_action :authenticate_user!
+  before_action -> { redirect_unless_user(:can_manage_committees?) }, only: [:new, :create]
+  before_action -> { redirect_unless_user(:can_edit_team?, team_from_params) }, only: [:edit]
+  before_action -> { redirect_unless_user(:can_edit_team?, team_from_params) }, only: [:update]
```
The two lines above should be done at once with `only: [:edit, :update]`

jonatanklosko commented on WcaOnRails/app/controllers/team_members_controller.rb:
```diff
@@ -0,0 +1,63 @@
+    else
+      render :edit
+    end
+  end
```
I think it makes sense redirect to team(s) instead of committee(s) in all actions above.

pingskills replied:
The committee page shows all the teams for that committee. For all but the delegates committee there is currently only 1 team. There isn't an individual team page as such to link to.

jonatanklosko replied:
Ah sure, thanks for explaining =)

jonatanklosko commented on WcaOnRails/app/controllers/teams_controller.rb:
```diff
@@ -1,54 +1,66 @@
-  end
+  before_action -> { redirect_unless_user(:can_manage_committees?) }, only: [:new, :create, :destroy]
+  before_action -> { redirect_unless_user(:can_edit_team?, team_from_params) }, only: [:edit]
+  before_action -> { redirect_unless_user(:can_edit_team?, team_from_params) }, only: [:update]
```
These two can be joined as mentioned before.

pingskills commented:
Thanks for the suggestions @jonatanklosko, I've updated the code.

jonatanklosko commented on WcaOnRails/app/helpers/static_pages_helper.rb:
```diff
@@ -1,8 +1,8 @@
-    Team.find_by_friendly_id!(team_friendly_id).current_members.includes(:user).order(team_leader: :desc).order("users.name asc").map do |u|
-      u.user.name + (u.team_leader ? " (leader)" : "")
+  def format_team_members(slug)
+    Team.find_by_slug!(slug).current_members.includes(:user).includes(:committee_position).order("committee_positions.team_leader desc").order("users.name asc").map do |u|
```
You can pass multiple arguments to `includes`, like `includes(:user, :committee_position)` =)

jonatanklosko commented on WcaOnRails/app/models/committee.rb:
```diff
@@ -0,0 +1,33 @@
+class Committee < ActiveRecord::Base
+  has_many :teams
+  has_many :team_members, through: :teams
+  has_many :committee_positions
```
I guess we want to delete committee positions if the committee is removed, so it makes sense to add `, dependent: :destroy` here.

pingskills replied:
The code doesn't show a delete button if the committee has a committee position or if it has teams but I think you're right, this should go here anyway.

jonatanklosko commented on WcaOnRails/app/models/committee_position.rb:
```diff
@@ -0,0 +1,25 @@
+  before_validation :compute_slug
+  private def compute_slug
+    self.slug ||= name.parameterize
+  end
```
Maybe we should benefit from `to_param` here as well?

pingskills replied:
I thought because no committee position is ever used in a URL it doesn't really need it here?

jonatanklosko replied:
Are you sure? What about [this](https://github.com/pingskills/worldcubeassociation.org/blob/f9f7f423ef6cdfe08a8d4e94958369b6c57f452e/WcaOnRails/app/controllers/committee_positions_controller.rb#L62-L64).

jonatanklosko commented:
**A point worth discussing:**
Do we want to create/update/delete teams and committees via UI?

**Concerns:**
Almost every team/committee is in our code (seeds/tests/permissions) so even if we perform the actions I mentioned, we still need to update our code and I think it could become messy to do it both ways.

jonatanklosko commented on WcaOnRails/app/views/committee_positions/_committee_position_form.html.erb:
```diff
@@ -0,0 +1,16 @@
+    <%= f.input :name %>
+    <%= f.input :description, hint: t('.description_html', md: t('common.hints.supports_md_html')) %>
+    <%= f.input :team_leader %>
+    <%= link_to icon("undo", "Back to #{@committee.name}"), committee_path(@committee), class: "btn btn-primary" %>
```
@viroulep what are your thoughts here? We use both translated hint that make sense at least because of reusing `'supports_md_html'`, and untranslated `"Back to #{@committee.name}"` that makes sense as well because it's not a public page. I think we don't want to end up with pages mixed of multiple languages.

viroulep replied:
Since the pages using this form are very small I'd suggest to localize the few remaining strings.
This we still feel a bit weird though, since the new/edit pages will be translated but not the rest. But it will probably be less weird than having different languages in the same form.

jonatanklosko commented on WcaOnRails/app/views/committee_positions/new.html.erb:
```diff
@@ -0,0 +1,8 @@
+<% provide(:title, 'Create a New Position In The ' + @committee.name) %>
+
+<div class="container new-team">
```
`new-team` is not accurate here, moreover I see that you've removed `.new-team` class so it doesn't make any sense to put it here =)

jonatanklosko commented on WcaOnRails/app/views/committees/show.html.erb:
```diff
@@ -0,0 +1,79 @@
+          </tr>
+        </thead>
+        <tbody>
+        <% team.current_members.includes(:user).includes(:committee_position).order("committee_positions.team_leader desc").order("users.name asc").each do |member| %>
```
`includes(:user, :committee_position)` =)

jonatanklosko commented on WcaOnRails/app/controllers/team_members_controller.rb:
```diff
@@ -0,0 +1,62 @@
+class TeamMembersController < ApplicationController
+  before_action :authenticate_user!
+  before_action -> { redirect_unless_user(:can_manage_committees?) }, only: [:new, :create]
+  before_action -> { redirect_unless_user(:can_edit_team?, team_from_params) }, only: [:edit, :update]
```
Shouldn't it just be `can_edit_team?` for all actions? If someone can manage team why wouldn't he add users? Also I think `can_manage_committees?` has nothing to do here.

pingskills replied:
Good points!

jonatanklosko commented on WcaOnRails/app/views/committees/show.html.erb:
```diff
@@ -0,0 +1,79 @@
+      <%= link_to icon("plus", "Create New Team"), new_committee_team_path(@committee), class: "btn btn-success" %>
+    <% end %>
+  </p>
+</div>
```
If I'm right that this page works as the public '/delegates' one, it should be internationalized.

jonatanklosko commented on WcaOnRails/app/views/layouts/_navigation.html.erb:
```diff
@@ -145,7 +145,7 @@
                 </li>
                 <% current_user.teams_where_is_leader.each do |team| %>
-                  <li><%= link_to team.name, edit_team_path(team) %></li>
+                  <li><%= link_to team.name, edit_committee_team_path(team.committee.slug, team.slug) %></li>
```
Let's make use of `to_param` method (that we have set up already) and just call `edit_committee_team_path(team.committee, team)`

viroulep commented on WcaOnRails/config/locales/fr.yml:
```diff
@@ -24,6 +24,8 @@ fr:
     wca_id: "ID WCA"
     all_regions: "Toutes les régions"
+    hints:
+      supports_md_html: ""
```
@pingskills, thanks for taking care of the French locale :)
You probably noticed that we have spec for i18n now, here is how it works:
- it will report any unused key in any language
- it will report any missing key in English locale

So I think you have to remove the keys you don't use (which you did), but you should probably not add any empty key for the new keys you introduced in the English locale: it's the burden of the maintainer, and spec-wise you're just fine since it will fallback to English and won't raise an error in travis :)
Also adding empty translation for ActiveRecords means French forms would have empty labels and be very weird ^^

For this particular key (`support_md_html`), could you please move the translation you removed from line 409 to here? It's would be the exact same ;)

pingskills replied:
Ah that makes sense. I'll remove the empty keys.

jonatanklosko commented on WcaOnRails/app/views/users/edit.html.erb:
```diff
@@ -114,17 +114,11 @@
             <label><%= t '.member_of' %></label>
             <% @user.current_teams.each do |team| %>
-              <%= link_to team.name, edit_team_path(team), class: "btn btn-xs btn-info #{"disabled" unless current_user.can_edit_team?(team)}" %>
+              <%= link_to team.name, edit_committee_team_path(team.committee.slug, team.slug), class: "btn btn-xs btn-info #{"disabled" unless current_user.can_edit_team?(team)}" %>
```
As before `edit_committee_team_path(team.committee, team)`

jonatanklosko commented on WcaOnRails/config/locales/en.yml:
```diff
@@ -125,6 +133,10 @@ en:
+      team_member:
+        end_date: "The date this member left the team"
+        start_date: "The date this member joined the team"
+        user_id: "team member"
```
Shouldn't it be capitalized?

jonatanklosko commented on WcaOnRails/db/migrate/20160731000001_create_committees.rb:
```diff
@@ -0,0 +1,14 @@
+  def change
+    create_table :committees do |t|
+      t.string :name, limit: 50, null: false
+      t.string :slug, limit: 50, null: false
```
Hmm, we use variable for limit validation, what if we make it larger in the future (just wondering).
@larspetrus what's the best practice, is it worth using SQL limit?

larspetrus replied:
I can't think of any good reason to have the SQL limit.

The validation is the right way to limit the size, and is easily changed if you want to.

There is some traditional belief that limiting the field saves disk space, but I don't think that is true in modern DBs, and even if it is, it would only matter if we had millions of Committees. And once we have that many committees, no work would get done anyway.

jonatanklosko replied:
@pingskills there are many migrations in this PR using limit and the structure.sql, please fix them =)

pingskills replied:
Yes millions of committees would certainly be a problem :)

As these are varchar fields they are variable length so there is no space saving by limiting the length. I think it's good practice though since it ensures database integrity. Is the rails application the only way we'll ever insert data? If we want to use a different tool to do reporting then having a sensible length set in the database helps understand the size of the data. If later on we want to extend the length this is easily done. To me it makes good sense to have limits in the database.

It is important to have consistency in projects so I have no issue in removing these limits as you suggested since that's how it is being done for all other migrations.

pingskills replied:
We did have `limit` in the `20151207230222_create_polls_tables.rb` migration.

jonatanklosko commented on WcaOnRails/db/migrate/20160731000002_insert_committees.rb:
```diff
@@ -0,0 +1,49 @@
+      'This team is responsible for managing the WCA''s software (website, scramblers, workbooks, admin tools).',
+      'software@worldcubeassociation.org', now(), now());
+    SQL
+  end
```
Creating those should be better with Rails model. When using SQL we don't fire callbacks (which may be added in the future). Also it makes things simpler as you don't have to use `NULL` and `now()`. Same for teams and positions.

That said note that we create all these things in seeds (with Rails models, yay) so I'd be up for single migration loading the seeds, simpler and not redundant

pingskills replied:
I remember above that it was suggested using SQL might be better since models can change over time?

jonatanklosko replied:
Hmm, actually I suggested to use seeds in in the migration to follow DRY.
Just about SQL vs Model I don't think using the first one is better, note that when we use SQL we omit validations and callbacks, which is not really good. Also one of the main reasons for having Models is to abstract queries to ruby code. Both approaches have cons and pros, so it may differ for other cases.

jonatanklosko commented on WcaOnRails/db/seeds/committees.seeds.rb:
```diff
@@ -0,0 +1,58 @@
+Committee.create(name: 'WCA Software Committee', \
+                 slug: Committee::WCA_SOFTWARE_COMMITTEE, \
+                 duties: software_duties, \
+                 email: 'software@worldcubeassociation.org')
```
We don't need all those `\`.
I would use JSON-like style, but that may be personal feeling and that's minor thing, so feel free to do whatever you like more.
```ruby
# Duties definition
...
Committee.craete [
  {
    name: 'WCA Software Committee',
    slug: Committee::WCA_SOFTWARE_COMMITTEE,
    duties: software_duties,
    email: 'software@worldcubeassociation.org'
  },
  {
    ....
  }
....
]
```

jonatanklosko commented on WcaOnRails/db/seeds/committees.seeds.rb:
```diff
@@ -0,0 +1,58 @@
+results_duties = <<EOS
+This committee is responsible for managing all WCA competition results.
+EOS
+@results = Committee.create(name: 'WCA Results Committee', \
```
Stray `@results` here.

jonatanklosko commented on WcaOnRails/db/seeds/development/competitions.seeds.rb:
```diff
@@ -11,7 +11,7 @@ def random_wca_value
   end
 
-  delegate = User.find_by(delegate_status: "delegate")
+  delegate = User.joins(:current_committees).where(committees: {slug: 'wca-delegates-committee'}).take
```
I think id could be `delegate = User.delegate`

pingskills replied:
Nice :)

jonatanklosko commented on WcaOnRails/db/seeds/development/users.seeds.rb:
```diff
@@ -13,37 +13,47 @@ def self.random_user
+  3.times do
+    FactoryGirl.create(:team_member, team: disciplinary_team)
   end
 end
```
Does it make sense to pass `team:` to the team_member factory? That's what the traits like `disciplinary_team_leader` takes care of.

jonatanklosko commented on WcaOnRails/spec/controllers/committee_positions_controller_spec.rb:
```diff
@@ -0,0 +1,180 @@
+  describe 'POST #create' do
+    context 'when not signed in' do
+      it 'redirects to the sign in page' do
+        post :create, committee_id: committee.slug, committee_position: {name: "An important position", description: "This is a very important position"}
```
You can use `committee_position: FactoryGirl.attributes_for(:committee_position)`
Also it makes sense to use `let(:committee_position_attributes)` as you do this in multiple places.

jonatanklosko commented on WcaOnRails/spec/controllers/committees_controller_spec.rb:
```diff
@@ -0,0 +1,221 @@
+        get :index
+        expect(response).to render_template :index
+      end
+    end
```
These two tests doesn't make much sense. If we can access something as non-signed-in then any user with any position can do so.

jonatanklosko replied:
But I guess you wanted to make all tests uniformed, so I don't have strong feeling for removing this if you like it =)

jonatanklosko commented on WcaOnRails/spec/controllers/committees_controller_spec.rb:
```diff
@@ -0,0 +1,221 @@
+        expect(response).to render_template :show
+      end
+    end
+  end
```
Same as above.

jonatanklosko commented on WcaOnRails/spec/controllers/committees_controller_spec.rb:
```diff
@@ -0,0 +1,221 @@
+      sign_out
+
+      it 'redirects to the sign in page' do
+        post :create, committee: { name: "Software", email: "software@worldcubeassociation.org", duties: "Responsible for all WCA software" }
```
See [this comment](https://github.com/cubing/worldcubeassociation.org/pull/725/files/#r77199556).

jonatanklosko commented on WcaOnRails/spec/controllers/competitions_controller_spec.rb:
```diff
@@ -463,20 +463,22 @@
 
   describe 'GET #post_announcement' do
+    let(:delegate_team_member) { FactoryGirl.create(:team_member, :senior_delegate, start_date: "2010-01-01") }
+    let(:competition_to_post) { FactoryGirl.create(:competition, delegates: [delegate_team_member.user]) }
```
You can use `FactoryGirl.create(:delegate, start_date: ...)` in the first line and omit `user` in the second one.

jonatanklosko commented on WcaOnRails/spec/controllers/team_members_controller_spec.rb:
```diff
@@ -0,0 +1,164 @@
+                      team_id: team_member_to_create.team.slug, \
+                      team_member: { user_id: team_member_to_create.user_id, \
+                                     start_date: team_member_to_create.start_date, \
+                                     end_date: team_member_to_create.end_date, \
```
Why do you use those `\`? They are not necessary =)

jonatanklosko commented on WcaOnRails/spec/controllers/team_members_controller_spec.rb:
```diff
@@ -0,0 +1,164 @@
+      it 'redirects to the sign in page' do
+        post :create, committee_id: team_member_to_create.committee.slug, \
+                      team_id: team_member_to_create.team.slug, \
+                      team_member: { user_id: team_member_to_create.user_id, \
```
This is really redundant you can use `team_member: team_member_to_create.attributes` I hope.

pingskills replied:
Much nicer :)

jonatanklosko commented on WcaOnRails/spec/controllers/teams_controller_spec.rb:
```diff
@@ -65,138 +47,128 @@
+      sign_in { FactoryGirl.create :admin_demoted }
+
+      it 'does not allow creation' do
+        post :create, committee_id: team.committee.slug, team: {name: "Team 2016", description: "An important team for 2016"}
```
See [this comment](https://github.com/cubing/worldcubeassociation.org/pull/725/files/#r77199556).

jonatanklosko commented on WcaOnRails/spec/controllers/teams_controller_spec.rb:
```diff
@@ -65,138 +47,128 @@
-      end
+      let(:leader) { FactoryGirl.create(:regulations_team_leader) }
+      let(:team_where_is_leader) { Team.find_by_slug('regulations-team') || FactoryGirl.create(:team, name: 'Regulations Team', committee: Committee.find_by_slug(Committee::WCA_REGULATIONS_COMMITTEE)) }
+      let(:team_where_is_not_leader) { Team.find_by_slug('software-team') || FactoryGirl.create(:team, name: 'Software Team', committee: Committee.find_by_slug(Committee::WCA_SOFTWARE_COMMITTEE)) }
```
1. You've been using constants in `Team.find_by_slug(...)` so far.
2. This `find || create` pattern is ugly, as a lot of other code relies on the current teams maybe it's worth seeding them the same way we do with countries.

jonatanklosko commented on WcaOnRails/spec/factories/team_members.rb:
```diff
@@ -1,9 +1,95 @@
+    trait :disciplinary_team_leader do
+      team do
+        Team.find_by_slug('disciplinary-team') || FactoryGirl.create(:team, name: 'Disciplinary Team', committee: Committee.find_by_slug(Committee::WCA_DISCIPLINARY_COMMITTEE))
+      end
```
This file doesn't use slug constants.
There's even more `find || create` pattern, seems worth having the teams in test db as countries.

jonatanklosko commented on WcaOnRails/spec/models/team_members_spec.rb:
```diff
@@ -0,0 +1,69 @@
+
+  it "ensures a user cannot have overlapping memberships to a single team" do
+    team_member = FactoryGirl.create :team_member
+    team_member2 = FactoryGirl.build :team_member, user_id: team_member.user_id, team_id: team_member.team_id, committee_position_id: team_member.committee_position_id
```
These two have *the same* dates, so the test doesn't test the actual overlapping, just add dates like from 4 Dec to 20 Dec and 15 Dec to now.

jonatanklosko commented on WcaOnRails/spec/models/team_members_spec.rb:
```diff
@@ -0,0 +1,69 @@
+  context "when adding members to a delegate team" do
+    it "ensures the first member added is a senior delegate" do
+      delegates_committee = Committee.find_by_slug(Committee::WCA_DELEGATES_COMMITTEE)
+      delegate_team = Team.where("slug = :slug", slug: 'delegate-testing-team', committee_id: delegates_committee.id).first_or_create(name: 'Delegate Testing Team', description: 'For testing', committee_id: delegates_committee.id)
```
We don't need `"slug = :slug"` here, `Team.where(slug: 'delegate-testing-team', ...)` should work perfectly fine.

`first_or_crate` is the same as `find || create` pattern mentioned before, so such place could be fixed as well we would use seeds.

pingskills replied:
I'll change this to use `let`

jonatanklosko commented:
Just went through the changes, firstly let's decide about [the point worth discussing](https://github.com/cubing/worldcubeassociation.org/pull/725#issuecomment-244053909), because if we resign from create/update/delete via UI then some code will be deleted and thus a part of my comments will just not apply.

pedrosino commented:
I think Jonatan's concerns are valid. When I first made the teams I thought of having a nice "+ Add team button", but now I see it can cause a lot of trouble.

jfly commented:
(note: I still have not looked at the code, but I do intend to soon) 

"committee" is already a term the WCA uses (WRC, WDC, etc), so I'm not wild about defining committee as "a collection of teams". I'm  especially loathe to invent vocabulary that conflicts with the vocabulary being used in the wca bylaws that were recently written. One idea would be to just have teams, and have a consistent naming scheme for all our delegate teams, so we can determine if someone is a delegate. Alternatively, I also like @larspetrus's idea of being able to nest teams. I'll have to look at the code to see what makes a committee different from a team that contains other teams.

As for Jonatan's comments about exactly what kind of CRUD ui we need, I agree with him that right now our teams are pretty tightly coupled with our code, and that there isn't much value in being able to create a team through the ui, as you'll still need to change our code. However, I'm guessing that with this pr, it would be possible to create a new region of delegates without having to go update our code to recognize members of this new team as delegates? Even if that's the case, we don't add/change/remove delegate regions often, so I would still prefer to keep things simple at first and not have a ui for this. We can always add one later.

pingskills commented:
All good points. I wanted to have a go at this change since I thought it would be good for me to learn more about the website and rails which I have certainly done. I have no issues if it doesn't quite fit and someone else wants to implement this in a cleaner, leaner manner.

jfly commented on WcaOnRails/app/views/committee_positions/new.html.erb:
```diff
@@ -0,0 +1,8 @@
+<% provide(:title, 'Create a New Position In The ' + @committee.name) %>
```
nit: In -> in

jfly commented on WcaOnRails/db/migrate/20160731000008_update_team_members_with_positions.rb:
```diff
@@ -0,0 +1,526 @@
+
+    execute "update team_members \
+             set    start_date = now() \
+             where  start_date = '0000-00-00';"
```
I've only been skimming everything, but could we use our knowledge of when competitions are and who delegated them to give more accurate start/end dates for our delegates?

pingskills replied:
That's what the SQL statement above does. Then if they haven't delegated a competition yet the start date is set to now().

jfly commented:
I just had a chat with @pingskills about this. Jeff is going to focus on #729 (payments). I'm going to close this PR for now. We'll reinvestigate delegate history tracking sometime later.
